tidy3d/plugins/design/design.py

```diff
@@ -86,38 +116,281 @@ def get_fn_source(function: Callable) -> str:
         except (TypeError, OSError):
             return None
 
-    def run(self, function: Callable, **kwargs) -> Result:
-        """Run the design problem on a user defined function of the design parameters.
+    def run(self, fn: Callable, fn_post: Callable = None, verbose: bool = True) -> Result:
+        """Explore a parameter space with a supplied method using the user supplied function.
+
+        Supplied functions are used to evaluate the design space and are called within the method.
+        For optimization methods these functions act as the fitness function. A single function can be
+        supplied which will contain the preprocessing, computation, and analysis of the desired problem.
+        If running a Tidy3D simulation, it is recommended to split this function into a pre function, that creates a Simulation object(s),
+        and a post function which analyses the SimulationData produced by the pre function Simulations. This allows the DesignSpace to
+        manage the batching of Simulations, which varies between Method used, and saving time writing their own batching code.
+        It also efficiently submits simulations to the cloud servers allowing for the fastest exploration of a design space.
+
+        The ``fn`` function must take a dictionary input - this can be stored a dictionary ``def example_fn(**params)``
+        or left as keyword arguments ``def example_fn(arg1, arg2)`` where the keywords correspond to the ``name`` of the parameters in the design space.
+
+        If used as a pre function, the output of ``fn`` must be a float, ``Simulation``, ``Batch``, list, or dict. Supplied ``Batch`` objects are
+        run without modification and are run in series. A list or dict of ``Simulation`` objects is flattened into a single ``Batch`` to enable
+        parallel computation on the cloud. The original structure is then restored for output; all `Simulation`` objects are replaced by ``SimulationData`` objects.
+
+        The output of ``fn_post`` (or ``fn`` if only one function is supplied) must be a float
+        or a container where the first element is a ``float`` and second element is a ``list`` / ``dict`` e,g. [float {"aux_1": str}].
+        The float is used by the optimizers as the return of the fitness function.
+        The second element is for auxiliary data from the analysis that the user may want to keep.
+        Sampling methods (``MethodGrid`` or ``MethodMonteCarlo``) can have any return type.
```

Comment:
I dont know if it's just my inability to read properly. but I do think just putting a markdown table might be helpful. For a given `fn_pre` output type, what operation(s) do we do behind the scene, and what type to expect as `fn_post` input.